### PR TITLE
[Fixes #2228] Use the config of a related cop whether it's enabled or not

### LINKED ISSUE #2228: --auto-gen-config not resolving all offenses after upgrading from 0.32.1
I'm having trouble reproducing this in an isolated code base, so here's the relevant information:

* 1779 ruby files
* many offenses listed in the `.rubocop_todo.yml` file, which was generated by rubocop 0.32.1
* everything is green in rubocop 0.32.1, but I'm trying to upgrade
* the `Lint/EndAlignment` cop is set to `AlignWith: variable`
* I have no `Lint/EndAlignment` offenses, but `Style/ElseAlignment` is disabled in the todo file
* I upgrade to a newer rubocop (both 0.33 and 0.34 have this behavior), and now I have some new offenses, because some new cops were introduced
* (my `.rubocop_todo.yml` is still the one generated by 0.32.1)
* I don't want to resolve all of the new offenses right now, I'd rather just regenerate the `.rubocop_todo.yml` file, so I run `bin/rubocop --auto-gen-config` and it regenerates my `.rubocop_todo.yml` file.

At this point, when I run `bin/rubocop`, there should be no offenses, right?

I start getting a bunch of `Style/ElseAlignment` offenses.

Before the regeneration (excerpts):

```yaml
# .rubocop_todo.yml
Style/ElseAlignment:
  Enabled: false

# .rubocop.yml
inherit_from .rubocop_todo.yml
Lint/EndAlignment:
  AlignWith: variable
```

After the regeneration (excerpts):

```yaml
# .rubocop_todo.yml
Lint/EndAlignment:
  Enabled: false

# .rubocop.yml
inherit_from .rubocop_todo.yml
Lint/EndAlignment:
  AlignWith: variable
```

Two interesting things:

1. `Lint/EndAlignment` is disabled even though it's got a configuration (new offenses seem to have been discovered in 0.33 related to the alignment of the `end` of case statements, but I'm not sure which entry in the changelog is responsible for this)
2. `Style/ElseAlignment` is no longer disabled, even though those offenses didn't go anywhere, so now I have a bunch of new offenses

Desired behavior: after running `bin/rubocop --auto-gen-config`, there would be no offenses

### PR DESCRIPTION
**Pull request by madwort**
I think this will fix #2228

### CODE REVIEW

**Comment by madwort** [comment]
(nb. needs tests!)

File: lib/rubocop/cop/mixin/space_after_punctuation.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@ def investigate(processed_source)
 
       def space_forbidden_before_rcurly?
         cfg = config.for_cop('Style/SpaceInsideBlockBraces')
-        style = cfg['Enabled'] ? cfg['EnforcedStyle'] : 'space'
+        style = cfg['EnforcedStyle'] ? cfg['EnforcedStyle'] : 'space'
```
Comment (alexdowad) [inline lib/rubocop/cop/mixin/space_after_punctuation.rb]: I haven't worked out the logic of this PR yet, but just regarding code style, this should be `style = cfg['EnforcedStyle'] || 'space'`. Likewise for the others below.

**Comment by bbatsov** [comment]
What exactly are you trying to achieve here? More info, please. :-)

**Comment by alexdowad** [comment]
He's trying to fix a problem where a codebase still has offenses after `--auto-gen-config`. Apparently, the problem is that when certain cops is disabled, their style is not set in quite the same way.

So say those cops are disabled when you run `--auto-gen-config`. The config which is generated for them  may not be what is actually needed to suppress offenses when the auto-genned config is used.

(I haven't tried running with this patch and seeing whether it actually fixes anything; this is just what I've gathered from reading GH posts.)

**Comment by madwort** [comment]
Yes, as @alexdowad the issue is:

These cops are configurable but they use the configuration of a different cop (I'm calling it "parent cop" for lack of a better term) instead of having their own configuration. For example, ElseAlignment uses the config of EndAlignment - which makes sense.
But they only use this other cop's config if the other cop is enabled, otherwise they use their own hard-coded default. 
All of these hard-coded defaults match the default configs in `default.yml`

This PR suggests changing the behaviour of rubocop config for these cops in the case that the default config for the parent cop is overriden by the user and the parent cop is also disabled. I think in all other cases, the config will be the same.
I think this will fix the problem noticed in #2228 when `--auto-gen-config` disabled the parent cop due to catching new violations introduced in newer versions of rubocop, which then caused the child cop to change its configuration from the user-configured value from the parent cop to the hard-coded default. 

If you think that this is better behaviour, then I will try & find time to fix up the tests (CI fails due to expecting the current behaviour, which I disagree with) & also add some more tests.

**Comment by alexdowad** [comment]
`--auto-gen-config` should generate a config file which suppresses all offenses. That is clear.

Personally I think that this business of "parent cops" is very problematic, and confusing for users. Each cop should have its own configuration. But since that is not the case right now, @madwort's fix is a good temporary solution to this trouble with `--auto-gen-config`.

**Comment by jonas054** [comment]
This change is a good idea IMO. There are some spec examples that set `Enabled: false` that are failing. These can be removed now.

**Comment by bbatsov** [comment]
The commits should be squashed and a changelog entry should be added.

**Comment by madwort** [comment]
Squashed & added a changelog entry

**Comment by bbatsov** [comment]
The travis build failed. I'd also suggest prefixing the commit message with `[Fix #2228] ` as this is our common convention.

**Comment by bbatsov** [comment]
You'll also have to rebase.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 ## master (unreleased)
 
+###Changes
+* [#2228](https://github.com/bbatsov/rubocop/issues/2228): Use the config of a parent cop whether Enabled or not ([@madwort][])
```
Comment (bbatsov) [inline CHANGELOG.md]: whether Enabled -> whether it's enabled

**Comment by bbatsov** [comment]
I'm also not sure about the term "parent cop" - "related cop" seems more appropriate to me.

**Comment by bbatsov** [comment]
ping

**Comment by madwort** [comment]
made the corrections but still need to check why travis fails.

**Comment by jonas054** [comment]
@madwort It's just the changelog entry that needs fixing. Add a dot at the end of the sentence, and link to your account at the bottom of the page.

**Comment by bbatsov** [comment]
And ideally you'd rebase again.

**Comment by madwort** [comment]
Revised changelog, rebased, pushed, CI is running.

**Comment by madwort** [comment]
Sorry for slow responses - been AFK for 10 days & then very busy on return!

**Comment by bbatsov** [comment]
No problem. :+1:

**Comment by madwort** [comment]
(also, thanks @jonas054 for the steer on fixing the changelog!)

**Pull request merged**